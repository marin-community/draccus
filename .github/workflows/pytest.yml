name: pytest

on: [push, pull_request]

jobs:
  test-on-env:
    strategy:
      matrix:
        os: [ubuntu-latest]
        python-version: ["3.9", "3.10", "3.11"]
    runs-on: ${{ matrix.os }}
    env:
      UV_PYTHON: ${{ matrix.python-version }}
    steps:
      - uses: actions/checkout@v2
      - name: set up UV with ${{ matrix.python-version }} on ${{ matrix.os }}
        uses: astral-sh/setup-uv@v6
      - name: Install dependencies
        run: |
          uv sync --locked --all-extras --dev
      - name: Test with pytest
        run: |
          uv run pytest
