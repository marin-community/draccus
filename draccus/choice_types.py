# SPDX-License-Identifier: MIT
# Copyright 2025 The Board of Trustees of the Leland Stanford Junior University

"""
A Choice Type aka "Sum Type" is a type that can be one of several types. Typically this is through subtyping,
but could be a tagged union or something else.

Typically you'll write a config like this:

```yaml
model:
    type: gpt
    pdrop: 0.1
    layers: 3
```

with a class hierarchy that looks like this:

```python
class ModelConfig(ChoiceRegistry):
    pass

@ModelConfig.register_subclass("gpt")
class GPTConfig(ModelConfig):
    pdrop: float
    layers: int
```

For now, Draccus choice types are implemented as classes that conform to a protocol. The protocol defines
a "get_choice_class(name)" method that returns the class corresponding to the given name, along with a few
other methods.

We may add support for Unions.

All registered types must be *dataclasses*
"""

import functools
import importlib
import pkgutil
from typing import Any, Callable, ClassVar, Dict, Optional, Protocol, Type, TypeVar, overload, runtime_checkable

from pydantic import TypeAdapter, model_serializer, model_validator

from draccus.utils import ParsingError

T = TypeVar("T")


CHOICE_TYPE_KEY = "type"
"""name of key to use in configuration to specify the type of a choice type"""


@runtime_checkable
class ChoiceType(Protocol):
    @classmethod
    def get_choice_class(cls, name: str) -> Any: ...

    @classmethod
    def get_known_choices(cls) -> Dict[str, Any]: ...

    @classmethod
    def get_choice_name(cls, subcls: Type) -> str:
        """
        Returns the name of the given subclass of this choice type.
        """
        ...

    @classmethod
    def default_choice_name(cls) -> Optional[str]:
        """
        Returns the name of the default subclass of this choice type.
        """
        ...


class ChoiceRegistryBase:
    _choice_registry: ClassVar[Dict[str, Any]]

    @model_validator(mode="wrap")
    @classmethod
    def validate_choice_registry(cls, values, handler) -> Any:
        if isinstance(values, cls):
            return values
        type_name = values.get(CHOICE_TYPE_KEY, None)
        if cls in cls.get_known_choices().values():
            # We are one of the choices already
            if type_name is not None:
                raise ParsingError(f"got choice key '{type_name}' when decoding a {cls} (as opposed to a superclass)")
            return handler(values)
        else:
            if type_name is None:
                type_name = cls.default_choice_name()
            if type_name is None:
                raise ParsingError(
                    f"expected a '{CHOICE_TYPE_KEY}' key with a value of one of {cls.get_known_choices().keys()}"
                )
            try:
                choice_class = cls.get_choice_class(type_name)
            except KeyError:
                raise ParsingError(
                    f"got choice key '{type_name}' but expected one of {cls.get_known_choices().keys()}"
                ) from None

            adapter = TypeAdapter(choice_class)
            return adapter.validate_python({k: v for k, v in values.items() if k != CHOICE_TYPE_KEY})

    @model_serializer(mode="wrap")
    def serialize_model(self, handler):
        result = handler(self)
        typ = self.__class__.get_choice_name(self.__class__)
        old_typ = result.setdefault(CHOICE_TYPE_KEY, typ)
        assert old_typ == typ, f"Expected {typ} but got {old_typ}"
        return result

    @classmethod
    def get_choice_class(cls, name: str) -> Any:
        return cls._choice_registry[name]

    @classmethod
    def get_known_choices(cls) -> Dict[str, Any]:
        return cls._choice_registry

    @classmethod
    def get_choice_name(cls, subcls: Type) -> str:
        for name, choice_type in cls._choice_registry.items():
            if choice_type == subcls:
                return name
        raise ValueError(f"Cannot find choice name for {subcls}")

    @classmethod
    def default_choice_name(cls) -> Optional[str]:
        return None

    @overload
    @classmethod
    def register_subclass(cls, name: str, choice_type: Type) -> Type[T]: ...

    @overload
    @classmethod
    def register_subclass(cls, name: str) -> Callable[[Type[T]], Type[T]]: ...

    @classmethod
    def register_subclass(cls, name: str, choice_type: Optional[Type[T]] = None):
        if choice_type is None:
            return functools.partial(cls.register_subclass, name)

        if name in cls._choice_registry:
            other_choice_type = cls._choice_registry[name]
            if other_choice_type != choice_type:
                raise ValueError(
                    f"Cannot register {choice_type} as {name} because {other_choice_type} is already registered as"
                    f" {name}"
                )

        cls._choice_registry[name] = choice_type
        return choice_type


class ChoiceRegistry(ChoiceRegistryBase):
    _choice_registry: ClassVar[Dict[str, Any]]

    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        if not hasattr(cls, "_choice_registry"):
            cls._choice_registry: ClassVar[Dict[str, Any]] = {}


class PluginRegistry(ChoiceRegistryBase):
    """
    A ChoiceRegistry that allows for discovery of plugins.

    It follows the "namespace package plugin" pattern described here:
    https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/#using-namespace-packages

    To use this, you should put all your plugins in a single package, and then pass the path to that package
    to the constructor of this class. Then, when the class looks for plugins, it will look in that package and
    automatically import them.

    This is useful for things like models, where you want to be able to add new models without having to modify
    the code of the library itself.

    Usage:
    ```python
    @dataclasses.dataclass
    class ModelConfig(PluginRegistry):

        pass

    @dataclasses.dataclass
    class GPTConfig(ModelConfig):
        pass

    ModelConfig.register_subclass("gpt", GPTConfig)

    the syntax for a yaml file would be:
    ```yaml
    model:
      type: gpt
      <config for gpt>
    ```

    Unlike with ClassRegistry, import doesn't happen until you call get_choice_class or get_known_choices,
    and you can split your plugins across multiple files.
    """

    # TODO: is it better to lazily import plugins?
    _choice_registry: ClassVar[Dict[str, Any]]
    discover_packages_path: ClassVar[str]
    _did_discover_packages: ClassVar[bool]

    def __init_subclass__(cls, discover_packages_path: Optional[str] = None, **kwargs):
        super().__init_subclass__(**kwargs)
        if not hasattr(cls, "_choice_registry"):
            cls._choice_registry = {}
        if not hasattr(cls, "discover_packages_path"):
            if discover_packages_path is None:
                raise ValueError("discover_packages_path must be specified in the class or constructor")
            cls.discover_packages_path = discover_packages_path
        cls._did_discover_packages = False

    @classmethod
    def get_choice_class(cls, name: str) -> Any:
        cls._discover_packages()
        return cls._choice_registry[name]

    @classmethod
    def get_known_choices(cls) -> Dict[str, Any]:
        cls._discover_packages()
        return cls._choice_registry

    @classmethod
    def _discover_packages(cls):
        if cls._did_discover_packages:
            return

        # from https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/
        package_module = importlib.import_module(cls.discover_packages_path, __package__)

        def iter_namespace(ns_pkg):
            # Specifying the second argument (prefix) to iter_modules makes the
            # returned name an absolute name instead of a relative one. This allows
            # import_module to work without having to do additional modification to
            # the name.
            return pkgutil.iter_modules(ns_pkg.__path__, ns_pkg.__name__ + ".")

        for _finder, pkg_name, _ispkg in iter_namespace(package_module):
            importlib.import_module(pkg_name)
            # registration should happen in the initialization of the package, so importing is sufficient

        cls._did_discover_packages = True
