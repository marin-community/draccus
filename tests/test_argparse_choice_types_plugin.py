# SPDX-License-Identifier: MIT
# Copyright 2025 The Board of Trustees of the Leland Stanford Junior University

import argparse
import dataclasses
import sys

import pytest

from draccus import ParsingError

from .draccus_choice_plugins.model_config import MlpConfig, ModelConfig
from .testutils import TestSetup


def test_plugin_registry_argparse():
    @dataclasses.dataclass(frozen=True)
    class Something(TestSetup):
        model: ModelConfig = MlpConfig(10, 5)

    s = Something.setup("")
    assert s.model == MlpConfig(10, 5)

    s = Something.setup("--model.type mlp --model.layers 12 --model.hidden_size 6")
    assert s.model == MlpConfig(12, 6)

    s = Something.setup("--model.layers 12 --model.hidden_size 6")
    assert s.model == MlpConfig(12, 6)

    s = Something.setup("--model.type gpt --model.layers 12 --model.attn_pdrop 0.2")
    from .draccus_choice_plugins.gpt import GptConfig

    assert s.model == GptConfig(12, 0.2)

    with pytest.raises(argparse.ArgumentError):
        Something.setup("--model.type baby")

    with pytest.raises(TypeError):
        Something.setup("--model.type gpt --model.layers 12 --model.hidden_size 6")

    with pytest.raises(TypeError):
        Something.setup("--model.attn_pdrop 12")


# skip this test if using python 3.8
# the help text is a bit different in 3.8


@pytest.mark.skipif(sys.version_info < (3, 10), reason="requires python3.9 or higher")
def test_choice_registry_examine_help(snapshot):
    @dataclasses.dataclass
    class Something(TestSetup):
        model: ModelConfig = MlpConfig(10, 5)

    assert Something.get_help_text() == snapshot
