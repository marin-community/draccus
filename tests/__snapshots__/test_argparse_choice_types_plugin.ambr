# serializer version: 1
# name: test_choice_registry_examine_help
  '''
  usage: draccus [-h] [--config_path str] [--model str] [--model.type {mlp,gpt}]
                 [--model.hidden_size int] [--model.layers int]
                 [--model.attn_pdrop float]
  
  options:
    -h, --help            show this help message and exit
    --config_path str     Path for a config file to parse with draccus (default:
                          None)
    --model str           Config file for model (default: None)
  
  test_choice_registry_examine_help.<locals>.Something:
  
  ModelConfig ['model']:
  
    --model.type {mlp,gpt}
                          Which type of ModelConfig ['model'] to use (default:
                          None)
  
  MlpConfig ['model']:
  
    --model.layers int
    --model.hidden_size int
  
  GptConfig ['model']:
  
    --model.layers int
    --model.attn_pdrop float
  '''
# ---
